### PR TITLE
Hci evt concat

### PR DESCRIPTION
**Pull request by michallowasrzechonek-silvair**

### CODE REVIEW

File: subsys/bluetooth/controller/hci/hci.c [hunk 1]
```diff
@@ -2141,146 +2149,192 @@ static inline bool dup_found(struct pdu_adv *adv)
...
-	rl_idx = b[offsetof(struct radio_pdu_node_rx, pdu_data) +
-		   offsetof(struct pdu_adv, payload) + adv->len + 1];
-	/* Update current RPA */
-	if (adv->tx_addr) {
-		ll_rl_crpa_set(0x00, NULL, rl_idx, &adv->adv_ind.addr[0]);
-	}
+   if(direct != 0)
+   {
+      return true;
+   }
```
Comment (michallowasrzechonek-silvair) [inline subsys/bluetooth/controller/hci/hci.c]: `return direct`

File: subsys/bluetooth/controller/hci/hci.c [hunk 2]
```diff
@@ -2141,146 +2149,192 @@ static inline bool dup_found(struct pdu_adv *adv)
...
 
-#if CONFIG_BT_CTLR_DUP_FILTER_LEN > 0
-	if (dup_found(adv)) {
-		return;
-	}
-#endif /* CONFIG_BT_CTLR_DUP_FILTER_LEN > 0 */
+   if (!(event_mask & BT_EVT_MASK_LE_META_EVENT) ||
+       !(le_event_mask & BT_EVT_MASK_LE_ADVERTISING_REPORT)) {
+      return;
+   }
```
Comment (michallowasrzechonek-silvair) [inline subsys/bluetooth/controller/hci/hci.c]: This should be checked inside the loop, after updating RPA

File: subsys/bluetooth/controller/hci/hci.c [hunk 3]
```diff
@@ -2141,146 +2149,192 @@ static inline bool dup_found(struct pdu_adv *adv)
...
+      if (adv->tx_addr) {
+         ll_rl_crpa_set(0x00, NULL, rl_idx, &adv->adv_ind.addr[0]);
+      }
+#endif
 
-		drp->num_reports = 1;
-		dir_info = (void *)(((u8_t *)drp) + sizeof(*drp));
+#if CONFIG_BT_CTLR_DUP_FILTER_LEN > 0
+      if (dup_found(adv)) {
+		   return;
```
Comment (michallowasrzechonek-silvair) [inline subsys/bluetooth/controller/hci/hci.c]: continue, not return

File: subsys/bluetooth/controller/hci/hci.c [hunk 4]
```diff
@@ -2141,146 +2149,192 @@ static inline bool dup_found(struct pdu_adv *adv)
...
-	adv_info = (void *)(((u8_t *)sep) + sizeof(*sep));
+   struct bt_hci_evt_le_direct_adv_report *drp;
+   struct bt_hci_evt_le_direct_adv_info *dir_info;
+   struct pdu_adv *adv = (void *)(node_rx->pdu_data);
 
-	adv_info->evt_type = c_adv_type[adv->type];
+   if (!(event_mask & BT_EVT_MASK_LE_META_EVENT) ||
+       !(le_event_mask & BT_HCI_EVT_LE_DIRECT_ADV_REPORT)) {
+      return;
+   }
```
Comment (michallowasrzechonek-silvair) [inline subsys/bluetooth/controller/hci/hci.c]: after updating RPA

File: subsys/bluetooth/controller/hci/hci.c [hunk 5]
```diff
@@ -2975,6 +3037,44 @@ void hci_evt_encode(struct radio_pdu_node_rx *node_rx, struct net_buf *buf)
 	}
 }
 
+u8_t hci_evt_concat(struct radio_pdu_node_rx **node_rx, struct net_buf *buf)
+{
+	u8_t node_cnt = 0;
+   struct radio_pdu_node_rx *node_rx_slice[HCI_MAX_NR_OF_CONCAT_MSG] = {NULL};
+
+	while(node_rx[node_cnt] && (node_cnt < HCI_MAX_NR_OF_CONCAT_MSG))
+	{
```
Comment (michallowasrzechonek-silvair) [inline subsys/bluetooth/controller/hci/hci.c:3053]: Add `node_rx_slice[node_cnt] = node_rx[node_cnt]` here, then you can remove the `memcpy` below

File: subsys/bluetooth/controller/hci/hci.c [hunk 6]
```diff
@@ -2975,6 +3037,44 @@ void hci_evt_encode(struct radio_pdu_node_rx *node_rx, struct net_buf *buf)
...
+		{
+			case NODE_RX_TYPE_REPORT:
+         {
+            if(is_direct_adv_report(node_rx[node_cnt])) {
+
+               /* Transmit previous packets - if exists */
+               if(node_cnt > 0) {
+                  duff_memcpy(node_rx_slice, node_rx, node_cnt * sizeof(struct radio_pdu_node_rx));
+                  le_advertising_report(node_rx_slice, buf);
+               }
```
Comment (michallowasrzechonek-silvair) [inline subsys/bluetooth/controller/hci/hci.c:3064]: `return` or at least `else`

File: subsys/bluetooth/controller/hci/hci_driver.c [hunk 7]
```diff
@@ -109,95 +109,132 @@ static void prio_recv_thread(void *p1, void *p2, void *p3)
...
 
 	/* process regular node from radio */
-	buf = encode_node(node_rx, class);
-
-	return buf;
+	return encode_node(node_rx, buf);
 }
 
 #if defined(CONFIG_BT_HCI_ACL_FLOW_CONTROL)
 static inline struct net_buf *process_hbuf(struct radio_pdu_node_rx *n)
```
Comment (michallowasrzechonek-silvair) [inline subsys/bluetooth/controller/hci/hci_driver.c:229]: Changes below shouldn't be needed anymore

File: subsys/usb/usb_device.c [hunk 8]
```diff
@@ -1357,6 +1357,15 @@ int usb_transfer_sync(u8_t ep, u8_t *data, size_t dlen, unsigned int flags)
...
 	k_sem_init(&pdata.sem, 0, 1);
 
+    /*
+	 * printk("USB data\r\n");
+	 * for(const uint8_t *i = &data[0]; i != &data[0] + dlen; i++)
+	 * {
+	 * 	printk("%x " , *i);
+	 * }
+	 * printk("#\r\n");
+     */
```
Comment (michallowasrzechonek-silvair) [inline subsys/usb/usb_device.c]: Remove

File: subsys/bluetooth/controller/hci/hci.c (hunk 2 above)
Comment (jakubwitowski-silvair) [reply subsys/bluetooth/controller/hci/hci.c]: Done

File: subsys/bluetooth/controller/hci/hci.c (hunk 3 above)
Comment (jakubwitowski-silvair) [reply subsys/bluetooth/controller/hci/hci.c]: Done

File: subsys/bluetooth/controller/hci/hci.c (hunk 4 above)
Comment (jakubwitowski-silvair) [reply subsys/bluetooth/controller/hci/hci.c]: Done

File: subsys/bluetooth/controller/hci/hci.c (hunk 1 above)
Comment (jakubwitowski-silvair) [reply subsys/bluetooth/controller/hci/hci.c]: Done

File: subsys/bluetooth/controller/hci/hci_driver.c [hunk 9]
```diff
@@ -152,7 +151,7 @@ static inline u8_t encode_node(struct radio_pdu_node_rx **node_rx, struct net_bu
     LL_ASSERT(!*buf);
 
     /* Only if all nodes are discardable */
-    if((*buf = bt_buf_get_rx(BT_BUF_EVT, 5))) // K_NO_WAIT)))
+    if((*buf = bt_buf_get_rx(BT_BUF_EVT, K_MSEC(5))))
```
Comment (michallowasrzechonek-silvair) [inline subsys/bluetooth/controller/hci/hci_driver.c:154]: :+1:

File: subsys/bluetooth/controller/hci/hci.c [hunk 10]
```diff
@@ -2975,6 +3050,58 @@ void hci_evt_encode(struct radio_pdu_node_rx *node_rx, struct net_buf *buf)
...
+            return 1;
+            break;
+         }
+		}
+
+		node_rx_slice[node_cnt] = node_rx[node_cnt];
+		node_cnt++;
+	}
+
+	le_advertising_report(node_rx, buf);
```
Comment (michallowasrzechonek-silvair) [inline subsys/bluetooth/controller/hci/hci.c]: I'd rather change the argument to `node_rx_slice`, and `goto` here from lines 3065 & 3082, avoiding copy pasta.

File: subsys/bluetooth/controller/hci/hci_driver.c [hunk 11]
```diff
@@ -307,23 +345,44 @@ static void recv_thread(void *p1, void *p2, void *p3)
...
 		BT_DBG("unblocked");
 
-		if (node_rx && !buf) {
-			/* process regular node from radio */
-			buf = process_node(node_rx);
+        if (node_rx[0] && !buf) {
+			/* process regular nodes from radio */
+            for (u8_t i = 0; (i < node_cnt) && node_rx[i]; i += process_node(&node_rx[i], &buf))
+            {
+                if (buf) {
```
Comment (michallowasrzechonek-silvair) [inline subsys/bluetooth/controller/hci/hci_driver.c]: I wonder if there is a cleaner way to remove copy pasta wrt lines 388-396

**Pull request merged**